manta-trusted-setup/src/groth16/test/bn254/ppot.rs

```diff
@@ -0,0 +1,1125 @@
+// Copyright 2019-2022 Manta Network.
+// This file is part of manta-rs.
+//
+// manta-rs is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// manta-rs is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with manta-rs.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Utility functions for Perpetual Powers of Tau
+
+use crate::{
+    groth16::kzg::{
+        Accumulator, Configuration as KzgConfiguration, G1Marker, G2Marker, Proof as KzgProof, Size,
+    },
+    ratio::RatioProof,
+    util::{BlakeHasher, Deserializer, KZGBlakeHasher, Serializer},
+};
+use ark_bn254::{Bn254, Fr, G1Affine, G2Affine, Parameters};
+use ark_std::{io, println};
+use blake2::Digest;
+use core::{fmt, marker::PhantomData};
+use manta_crypto::arkworks::{
+    ec::{
+        models::{bn::BnParameters, ModelParameters},
+        short_weierstrass_jacobian::GroupAffine,
+        AffineCurve, PairingEngine, SWModelParameters,
+    },
+    ff::{PrimeField, ToBytes, Zero},
+    pairing::Pairing,
+    serialize::{CanonicalSerialize, Read, SerializationError, Write},
+};
+use manta_util::{cfg_iter, into_array_unchecked, vec::Vec};
+use memmap::Mmap;
+
+#[cfg(feature = "rayon")]
+use manta_util::rayon::prelude::ParallelIterator;
+
+#[derive(Debug, PartialEq, Eq)]
+/// Configuration of the Perpetual Powers of Tau ceremony
+pub struct PerpetualPowersOfTauCeremony<S, const POWERS: usize> {
+    __: PhantomData<S>,
+}
```

Comment:
A few things to keep in mind:

1. Put the docstring on the first line.
2. Because the Rust `derive` macro generates the wrong semantics, use `derivative` and since this is just a `PhantomData` marker struct, implement all the standard traits (see below) in alphabetical order.
3. Since it's only a marker struct, use the tuple-struct form of the constructor `Struct(_)` rather than `Struct { .. }`.

```suggestion
/// Configuration of the Perpetual Powers of Tau Ceremony
#[derive(derivative::Derivative)]
#[derivative(Clone, Copy, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
pub struct PerpetualPowersOfTauCeremony<S, const POWERS: usize>(PhantomData<S>);
```